File: Graphs/695-max_area_of_island/695.c



int directions[4][2] = {{0, 1}, {0, -1}, {1, 0}, {-1, 0}};

int dfs(int x, int y, int** grid, int rows, int cols) {
    grid[x][y] = 0;
    int a;
    int b;
    int area = 1;
    for (int i = 0; i < 4; i++) {
        a = x + directions[i][0];
        b = y + directions[i][1];
        if ((a >= 0) && (b >= 0) && (a < rows) && (b < cols) && (grid[a][b] == 1)) {
            area += dfs(a, b, grid, rows, cols);
        }
    }
    return area;
}

int max(int a, int b) {
    if (a > b) {
        return a;
    }
    return b;
}

int maxAreaOfIsland(int** grid, int gridSize, int* gridColSize){
    
    int rows = gridSize;
    int cols = gridColSize[0];
    int max_area = 0;
    
    for (int row = 0; row < rows; row++) {
        for (int col = 0; col < cols; col++) {
            if (grid[row][col] == 1) {
                max_area = max(max_area, dfs(row, col, grid, rows, cols));
            }
        }
    }
    return max_area;
}

File: Graphs/695-max_area_of_island/695.rs

use std::cmp::max;

impl Solution {
    
    pub fn max_area_of_island(grid: Vec<Vec<i32>>) -> i32 {
        let rows: usize = grid.len();
        let cols: usize = grid[0].len();
        let mut max_area: usize = 0;
        let mut grid = grid;
        
        fn dfs(x: usize, y: usize, grid: &mut Vec<Vec<i32>>) -> usize {
            grid[x][y] = 0;
            let mut area: usize = 1;
            if x + 1 < grid.len() && grid[x + 1][y] == 1 {
                area += dfs(x + 1, y, grid);
            } 
            if x > 0 && grid[x - 1][y] == 1 {
                area += dfs(x - 1, y, grid);
            }
            if y + 1 < grid[0].len() && grid[x][y + 1] == 1 {
                area += dfs(x, y + 1, grid);
            }
            if y > 0 && grid[x][y - 1] == 1 {
                area += dfs(x, y - 1, grid);
            } 
            return area;
        }
        
        for row in 0..rows {
            for col in 0..cols {
                if grid[row][col] == 1 {
                    max_area = max(max_area, dfs(row, col, &mut grid));
                }
            }
        }
        return max_area as i32;
    }
}